File: dart/lib/src/twitter/change_verified_pubkey.dart
import '../constants/addresses.dart';
import '../instructions/delete_name_registry_instruction.dart';
import '../instructions/instruction_types.dart';
import '../instructions/transfer_instruction.dart';
import '../rpc/rpc_client.dart';
import '../utils/get_domain_key_sync.dart';
import 'create_reverse_twitter_registry.dart';

/// Change verified public key instruction
///
/// This function mirrors js/src/twitter/changeVerifiedPubkey.ts
///
/// Change the verified pubkey for a given twitter handle
/// Signed by the Authority, the verified pubkey and the payer
///
/// [connection] - RPC connection
/// [twitterHandle] - The Twitter handle
/// [currentVerifiedPubkey] - The current verified public key
/// [newVerifiedPubkey] - The new verified public key
/// [payerKey] - The payer's public key
///
/// Returns list of transaction instructions
Future<List<TransactionInstruction>> changeVerifiedPubkey(
  RpcClient connection,
  String twitterHandle,
  String currentVerifiedPubkey,
  String newVerifiedPubkey,
  String payerKey,
) async {
  final hashedTwitterHandle = getHashedNameSync(twitterHandle);
  final twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    nameParent: twitterRootParentRegistryAddress,
  );

  final instructions = <TransactionInstruction>[];

  // Transfer the user-facing registry ownership
  final transferInstr = TransferInstruction(
    newOwner: newVerifiedPubkey,
    params: TransferInstructionParams(
      newOwner: newVerifiedPubkey,
      programAddress: nameProgramAddress,
      domainAddress: await twitterHandleRegistryKey,
      currentOwner: currentVerifiedPubkey,
    ),
  );
  instructions.add(transferInstr.build());

  // Delete the old reverse registry
  final deleteInstr = DeleteNameRegistryInstruction();
  instructions.add(deleteInstr.getInstruction(
    programAddress: nameProgramAddress,
    domainAddress: await _getReverseRegistryKey(currentVerifiedPubkey),
    refundTarget: payerKey,
    owner: twitterVerificationAuthority,
  ));

  // Create the new reverse registry
  final reverseInstructions = await createReverseTwitterRegistry(
    connection,
    twitterHandle,
    await twitterHandleRegistryKey,
    newVerifiedPubkey,
    payerKey,
  );
  instructions.addAll(reverseInstructions);

  return instructions;
}

/// Get reverse registry key for a verified pubkey
Future<String> _getReverseRegistryKey(String verifiedPubkey) async {
  final hashedVerifiedPubkey = getHashedNameSync(verifiedPubkey);
  return getNameAccountKeySync(
    hashedVerifiedPubkey,
    nameClass: twitterVerificationAuthority,
    nameParent: twitterRootParentRegistryAddress,
  );
}

File: js/src/twitter/changeVerifiedPubkey.ts
import { Connection, PublicKey, TransactionInstruction } from "@solana/web3.js";
import {
  NAME_PROGRAM_ID,
  TWITTER_VERIFICATION_AUTHORITY,
  TWITTER_ROOT_PARENT_REGISTRY_KEY,
} from "../constants";
import { deleteNameRegistry } from "../bindings/deleteNameRegistry";
import { transferInstruction } from "../instructions/transferInstruction";
import { getHashedNameSync } from "../utils/getHashedNameSync";
import { getNameAccountKeySync } from "../utils/getNameAccountKeySync";
import { createReverseTwitterRegistry } from "./createReverseTwitterRegistry";

/**
 * Builds instructions to transfer a Twitter handle to a new verified public key.
 *
 * The authority, current verified key, and payer must sign the resulting instructions.
 *
 * @param connection Solana RPC connection used to calculate rent-exemption costs.
 * @param twitterHandle Verified Twitter handle to transfer.
 * @param currentVerifiedPubkey Current verified owner and required signer.
 * @param newVerifiedPubkey Public key that will become the verified owner.
 * @param payerKey Signer that funds creation of the new reverse registry.
 * @returns Instructions that transfer the handle and recreate its reverse registry.
 *
 * @example
 * ```ts
 * const instructions = await changeVerifiedPubkey(
 *   connection,
 *   "bonfida",
 *   currentOwner,
 *   newOwner,
 *   payer,
 * );
 * ```
 */
export async function changeVerifiedPubkey(
  connection: Connection,
  twitterHandle: string,
  currentVerifiedPubkey: PublicKey,
  newVerifiedPubkey: PublicKey,
  payerKey: PublicKey,
): Promise<TransactionInstruction[]> {
  const hashedTwitterHandle = getHashedNameSync(twitterHandle);
  const twitterHandleRegistryKey = getNameAccountKeySync(
    hashedTwitterHandle,
    undefined,
    TWITTER_ROOT_PARENT_REGISTRY_KEY,
  );

  // Transfer the user-facing registry ownership
  let instructions = [
    transferInstruction(
      NAME_PROGRAM_ID,
      twitterHandleRegistryKey,
      newVerifiedPubkey,
      currentVerifiedPubkey,
      undefined,
    ),
  ];

  instructions.push(
    await deleteNameRegistry(
      connection,
      currentVerifiedPubkey.toString(),
      payerKey,
      TWITTER_VERIFICATION_AUTHORITY,
      TWITTER_ROOT_PARENT_REGISTRY_KEY,
    ),
  );

  // Create the new reverse registry
  instructions = instructions.concat(
    await createReverseTwitterRegistry(
      connection,
      twitterHandle,
      twitterHandleRegistryKey,
      newVerifiedPubkey,
      payerKey,
    ),
  );

  return instructions;
}
